hudi-common/src/main/java/org/apache/hudi/metadata/SecondaryIndexRecordGenerationUtils.java

```diff
@@ -0,0 +1,376 @@
+    try {
+      if (record.toIndexedRecord(tableSchema, CollectionUtils.emptyProps()).isPresent()) {
+        GenericRecord genericRecord = (GenericRecord) (record.toIndexedRecord(tableSchema, CollectionUtils.emptyProps()).get()).getData();
+        String secondaryKeyFields = String.join(".", indexDefinition.getSourceFields());
+        return HoodieAvroUtils.getNestedFieldValAsString(genericRecord, secondaryKeyFields, true, false);
+      }
+    } catch (IOException e) {
+      LOG.debug("Failed to fetch secondary key for record key " + record.getKey().toString());
+    }
+    return null;
+  }
+
+  public static HoodieData<HoodieRecord> readSecondaryKeysFromFileSlices(HoodieEngineContext engineContext,
+                                                                         List<Pair<String, FileSlice>> partitionFileSlicePairs,
+                                                                         int secondaryIndexMaxParallelism,
+                                                                         String activeModule, HoodieTableMetaClient metaClient, EngineType engineType,
+                                                                         HoodieIndexDefinition indexDefinition) {
+    if (partitionFileSlicePairs.isEmpty()) {
+      return engineContext.emptyHoodieData();
+    }
+    final int parallelism = Math.min(partitionFileSlicePairs.size(), secondaryIndexMaxParallelism);
+    final StoragePath basePath = metaClient.getBasePath();
+    Schema tableSchema;
+    try {
+      tableSchema = new TableSchemaResolver(metaClient).getTableAvroSchema();
+    } catch (Exception e) {
+      throw new HoodieException("Failed to get latest schema for " + metaClient.getBasePath(), e);
+    }
+
+    engineContext.setJobStatus(activeModule, "Secondary Index: reading secondary keys from " + partitionFileSlicePairs.size() + " file slices");
+    return engineContext.parallelize(partitionFileSlicePairs, parallelism).flatMap(partitionAndBaseFile -> {
+      final String partition = partitionAndBaseFile.getKey();
+      final FileSlice fileSlice = partitionAndBaseFile.getValue();
+      List<String> logFilePaths = fileSlice.getLogFiles().sorted(HoodieLogFile.getLogFileComparator()).map(l -> l.getPath().toString()).collect(Collectors.toList());
+      Option<StoragePath> dataFilePath = Option.ofNullable(fileSlice.getBaseFile().map(baseFile -> filePath(basePath, partition, baseFile.getFileName())).orElseGet(null));
+      Schema readerSchema;
+      if (dataFilePath.isPresent()) {
+        readerSchema = HoodieIOFactory.getIOFactory(metaClient.getStorage())
+            .getFileFormatUtils(metaClient.getTableConfig().getBaseFileFormat())
+            .readAvroSchema(metaClient.getStorage(), dataFilePath.get());
+      } else {
+        readerSchema = tableSchema;
+      }
+      return createSecondaryIndexGenerator(metaClient, engineType, logFilePaths, readerSchema, partition, dataFilePath, indexDefinition,
```

Comment:
if the previous slice only contains base file, we should consider directly reading it via HoodieIOFactory and parquet utils way. 
again, we can do that in a follow up PR

Reply:
can you help remind me, what happens when ingestion commit (DC), adds a new base file for an existing file slice? 
does prev file slice contain entire file slice (base file + all log files) ? 
do we have tests for this scenario.

Reply:
> what happens when ingestion commit (DC), adds a new base file for an existing file slice?

I guess you mean filegroup and not file slice, because a file slice is associated with one base file right. So, in that case previous file slice will contain the base file and the log files only as of the last committed time. While the current file slice will contain new base file.

> do we have tests for this scenario.

Doesn't COW basically cover this?

Reply:
As discussed offline, bypassed the SI records generation for compaction/clustering.